### PR TITLE
Checkstyle: prevent upper case abbreviations in names #319

### PR DESCRIPTION
**Pull request by eugenepeh**
#319 
Ready for review

### CODE REVIEW

**Comment by CanIHasReview** [comment]
# v1

@eugenepeh submitted v1 for review.

* [v1 1/2] [Checkstyle: add AbbreviationAsWordInName rule](https://github.com/se-edu/addressbook-level4/pull/360/commits/c05b610e7468985aa40b0a6027eb9fd581143486)
* [v1 2/2] [checkstyle: fix AbbreviationAsWordInName rule violations ](https://github.com/se-edu/addressbook-level4/pull/360/commits/a0f0d967eafc8a3e1c86a5d7c40aba811097a482)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/27cb4621dca1bf474b2838352a90873f3fbcb4ef...a0f0d967eafc8a3e1c86a5d7c40aba811097a482))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/360/1/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

**Comment by CanIHasReview** [comment]
# v2

@eugenepeh submitted v2 for review.

* [v2 1/2] [Checkstyle: add AbbreviationAsWordInName rule](https://github.com/se-edu/addressbook-level4/pull/360/commits/c05b610e7468985aa40b0a6027eb9fd581143486)
* [v2 2/2] [Checkstyle: fix AbbreviationAsWordInName rule violations ](https://github.com/se-edu/addressbook-level4/pull/360/commits/b557e5186aaae9f61844ede89f930f8b17a1f22e)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/27cb4621dca1bf474b2838352a90873f3fbcb4ef...b557e5186aaae9f61844ede89f930f8b17a1f22e)) ([:chart_with_upwards_trend: Interdiff between v1 and v2](https://github.com/se-edu/addressbook-level4/compare/22c229b6c6622045530581be6e2f5b4c77273895~1...22c229b6c6622045530581be6e2f5b4c77273895))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/360/2/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

**Comment by CanIHasReview** [comment]
# v3

@eugenepeh submitted v3 for review.

* [v3 1/1] [Checkstyle: add AbbreviationAsWordInName rule and fix violations](https://github.com/se-edu/addressbook-level4/pull/360/commits/a5531464cedc65f2fdddc03f012481a92a21a515)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/27cb4621dca1bf474b2838352a90873f3fbcb4ef...a5531464cedc65f2fdddc03f012481a92a21a515)) ([:chart_with_upwards_trend: Interdiff between v2 and v3](https://github.com/se-edu/addressbook-level4/compare/8cc42b2f649e71b3ff4b8d17b4371c9646c3602b~1...8cc42b2f649e71b3ff4b8d17b4371c9646c3602b))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/360/3/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

File: config/checkstyle/checkstyle.xml [hunk 1]
```diff
@@ -81,6 +81,12 @@
     <!--
     NAMING CHECKS
     -->
+    <module name="AbbreviationAsWordInName">
+      <!-- Validate abbreviations(consecutive capital letters) length in identifier name -->
```
Comment (yamgent) [inline config/checkstyle/checkstyle.xml]: Needs a space in between for `abbrevations(consecutive)`

**Comment by CanIHasReview** [comment]
# v4

@eugenepeh submitted v4 for review.

* [v4 1/1] [Checkstyle: add AbbreviationAsWordInName rule and fix violations](https://github.com/se-edu/addressbook-level4/pull/360/commits/14f55073197dcb2b8e5138e9b43464b49ee27506)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/27cb4621dca1bf474b2838352a90873f3fbcb4ef...14f55073197dcb2b8e5138e9b43464b49ee27506)) ([:chart_with_upwards_trend: Interdiff between v3 and v4](https://github.com/se-edu/addressbook-level4/compare/bc884fb1054473ddbbd7c912d3a7ca03fcefa71a~1...bc884fb1054473ddbbd7c912d3a7ca03fcefa71a))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/360/4/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

**Comment by CanIHasReview** [comment]
# v5

@eugenepeh submitted v5 for review.

* [v5 1/1] [Checkstyle: add AbbreviationAsWordInName rule](https://github.com/se-edu/addressbook-level4/pull/360/commits/e6c3e50d926e6109559b99f90195992ca1548b9e)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/27cb4621dca1bf474b2838352a90873f3fbcb4ef...e6c3e50d926e6109559b99f90195992ca1548b9e)) ([:chart_with_upwards_trend: Interdiff between v4 and v5](https://github.com/se-edu/addressbook-level4/compare/9e3dd3c3f2eb534bb5a2ae2081eed0d865bceecf~1...9e3dd3c3f2eb534bb5a2ae2081eed0d865bceecf))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/360/5/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

File: config/checkstyle/checkstyle.xml [hunk 2]
```diff
@@ -81,6 +81,12 @@
     <!--
     NAMING CHECKS
     -->
+    <module name="AbbreviationAsWordInName">
+      <!-- Validate abbreviations (consecutive capital letters) length in identifier name -->
+      <property name="ignoreFinal" value="false"/>
+      <property name="allowedAbbreviationLength" value="1"/>
+      <property name="severity" value="error"/>
```
Comment (PierceAndy) [inline config/checkstyle/checkstyle.xml]: As mentioned elsewhere, this line can be omitted.

**Comment by CanIHasReview** [comment]
# v6

@eugenepeh submitted v6 for review.

* [v6 1/1] [Checkstyle: add AbbreviationAsWordInName rule](https://github.com/se-edu/addressbook-level4/pull/360/commits/6e00a6ad9651c354bfd2cf2e61730c7eb9107e73)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/a61db1bf27f62cc549eb79be0a4598f2f1994b9b...6e00a6ad9651c354bfd2cf2e61730c7eb9107e73)) ([:chart_with_upwards_trend: Interdiff between v5 and v6](https://github.com/se-edu/addressbook-level4/compare/8b4a8b70d623997afe37e186a468e9c27aa755b7~1...8b4a8b70d623997afe37e186a468e9c27aa755b7))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/360/6/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

**Comment by CanIHasReview** [comment]
# v7

@eugenepeh submitted v7 for review.

* [v7 1/1] [Checkstyle: add AbbreviationAsWordInName rule](https://github.com/se-edu/addressbook-level4/pull/360/commits/028ec75dec5dcaa4bb5be91afcf2ca8fec561ab0)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/07367fab5054768f3243fb55eba1367d3d67a39b...028ec75dec5dcaa4bb5be91afcf2ca8fec561ab0)) ([:chart_with_upwards_trend: Interdiff between v6 and v7](https://github.com/se-edu/addressbook-level4/compare/48545a472b182122e9f932c071c0d75955358145~1...48545a472b182122e9f932c071c0d75955358145))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/360/7/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

**Comment by damithc** [comment]
What do we do for a case such as `canIHasReview`? Will checkstyle reject it?

**Comment by PierceAndy** [comment]
It would. The problem would occur at word boundaries involving single-lettered words.
There are two classes of words that would trigger it:
1. Proper English words, i.e. `A` as in `handleAClass(...)` or `I` as in `CanIHasReview`.
2. Single letters, e.g. `B` in `PlanBFactory`.

These words are unlikely to occur in identifier names, but not impossible, potentially leading to cases with 2 consecutive uppercase letters. A workaround is having the offending names renamed, but that imposes an arbitrary restriction on naming.

Grammatically, it is **highly** unlikely for 2 of such words to occur consecutively, which would potentially lead to cases with 3 consecutive uppercase letters.

Therefore, I think it would be safer to set the value to 2.

**Comment by eugenepeh** [comment]
True but we wouldn't want legit 2 letters abbreviation be left unnoticed too,
perhaps I can set warning for the value 1 (more than 1 consecutive uppercase) and error for the value 2?

**Comment by damithc** [comment]
> True but we wouldn't want legit 2 letters abbreviation be left unnoticed too,

Indeed, one of the common violations is `PersonID`. 

Given that `canIHasReview` sort of thing is rare, we can reject those by default but fallback on one of these when we 'die die must use' such a name:
1. use checkstyle annotations skip the check
2. update the checkstyle regex to specifically allow the approved name

Your thoughts?

**Comment by eugenepeh** [comment]
I think stricter enforcement is good as reviewers may miss out such violation. Also, just like Andy mentioned, these words are unlikely to occur in identifier names and can always renamed.

There is also a existing rule in our checkstyle, [SuppressionCommentFilter](http://checkstyle.sourceforge.net/config_filters.html#SuppressionCommentFilter), that can be used to bypass checks if the reason of 'die die must use' such a name ever justifies the use of such a bypass.

**Comment by yamgent** [comment]
> Using allowedAbbreviations property is recommended in this case instead, as recurrence of the abbreviation violation variable/method will all require bypass otherwise.

That is a good idea. Should we document this finding somewhere?

**Comment by damithc** [comment]
> Should we document this finding somewhere?

In the commit message?

**Comment by CanIHasReview** [comment]
# v8

@eugenepeh submitted v8 for review.

* [v8 1/1] [Checkstyle: add AbbreviationAsWordInName rule](https://github.com/se-edu/addressbook-level4/pull/360/commits/c319ba530aaa14dc395f1656cfada7652d820bb2)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/6951a73c502185c54370c16fde85605743899d55...c319ba530aaa14dc395f1656cfada7652d820bb2)) ([:chart_with_upwards_trend: Interdiff between v7 and v8](https://github.com/se-edu/addressbook-level4/compare/0a3e4a5d330314d15f4e494de91977045d3b7391~1...0a3e4a5d330314d15f4e494de91977045d3b7391))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/360/8/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

**Comment by damithc** [comment]
I'll set this to `toMerge`. I'll refine the commit message based on @PierceAndy's comments when I merge.

**Comment by damithc** [comment]
Merged with a slightly revised commit message.

**Pull request merged**